### ISSUE TITLE
New damage system

### ISSUE THREAD

**Issue by MarkuBu** [minetest/minetest#6773]
The current damage system doesn't really work well, especially if armor is added. I set up a simple subgame with mobs redo, some default mobs and 3d armor.

Without wearing any armor even weak mobs like the sand monster are strong (and they should). But just with a wooden armor (without shield) sand monster doesn't make any damage. You can stand within 20 mobs and nothing happens. 

You need to add stronger mobs, but they are absolutely deadly as long as you don't have a full diamond armor. 

Attacks should always make damage, just the power of the attack should be reduced by the armor. Even a sand monster should make damage when wearing a diamond armor. 

I know that it is possible to make a custom damage system, but even with the KISS philosophy of Minetest it should have a working damage system. 

I suggest to change the current damage system (or optionally add a new one and leave the current one for backward compatibility). Health can still have a value of 20, but should be a decimal number (also other values like breath) and the maximum value should be settable by mods, even in a running game. Damage should be calculated as fraction instead of just reducing the attack power by the armor value (implementation details can be discussed in this issue).

Armor should also get wear when it get's damage

**Comment by ThomasMonroe314** [comment]
that is a very good idea, but would you be willing to make a PR for example purposes?

**Comment by MarkuBu** [comment]
All I can do is to write a lua mod. I don't know enough about the MT core. But even in Lua I have to look how it works. Will take some time

**Comment by jastevenson303** [comment]
>Armor should also get wear when it get's damage

https://github.com/jastevenson303/minetest-3d_armor/blob/dcbl/3d_armor/api.lua#L365

**Comment by MarkuBu** [comment]
I know, but this is done in a mod, not in the engine. Don't know if it makes sense, but I want to mention it

**Comment by Ferk** [comment]
There's no concept of "armor equipment" for the engine. There's only armor groups assigned to entities. And these already operate with percentages that can be very precise. The big jump in protection between no armor and wood or diamond is not something decided by the engine, and imho, it shouldn't be.

The problems you are encountered are related to the combination of the armor mod and mobs mod that you might be using. Maybe you should raise the issue in the github for either of those mods.

**Comment by MarkuBu** [comment]
I know that the engine has currently no armor "equipment". And I didn't know that the damage system works with percentage. Is health internally a float?

**Comment by Ferk** [comment]
This is the code in C that calculates the damage:
https://github.com/minetest/minetest/blob/c252ed506eb8324fbdea4b632116c2c7d06a8c9c/src/tool.cpp#L240-L250

The damage is still an integer, but the operations to increment it use floats (and you can see the constribution of the armor group is in percentage ``* armor / 100.0``), I guess there's an implicit casting (shouldn't this give a warning, though?).

**Comment by MarkuBu** [comment]
To be honest, I don't really understand that forular. And I really tried it

I broke down the example from the wiki page https://dev.minetest.net/Damage_system to a simple formular. As far as I understand an armor is better the lower the fleshy value in the armor_group is.

broken down to the minimum this should be the formular for the example (assuming actual_interval is 0.8)

`damage = 8 * 100.0 / 100.0`

With a lower armor group value (fleshy=20) I get lower damage

`damage = 8 * 20.0 / 100.0 = 1.6`

But the armor_group values in 3d armor are higher the better the armor is. That's what I don't understand. Or did I made a mistake somewhere?

The damage system is really complicated and I admit that I didn't understand it completely

**Comment by Ferk** [comment]
> But the armor_group values in 3d armor are higher the better the armor is. That's what I don't understand.

Yes, I'm also not familiar with how the 3d armor mod really works. I assume it modifies the armor group of the player, but it's not clear to me in which way.

From the looks of it, they added some sort of "elemental" protection (there's some ``armor_"..element`` stuff in the code). I think you should raise the issue in the git of the armor mod. Maybe it turns out that the behavior is intended and the wooden armor is designed to be extremely effective against sand monsters.

**Comment by CloudyProton** [comment]
When using mobs_monsters in conjunction with 3d_armor, I have found it a good idea to double the damage output of each hostile mob. A bit of a temporary fix until either the mods get adjusted or a damage system change, as you say.

**Comment by MarkuBu** [comment]
> I'm also not familiar with how the 3d armor mod really works.

Probably they subtract the values from the armor from the fleshy armor_group from the player. A sand monster makes only 1 damage, so the formula would look like this

`damage = 1 * (100.0 - 5 - 10 - 10 - 5) / 100.0 = 0.8`

And probably because it is only integer the decimal places are just cut off

> I have found it a good idea to double the damage output of each hostile mob.

That's no good idea because these mobs are deadly if you have no armor, for example after you died. If you spawn near some mobs you get instantly killed and you can delete the game and start a new one because you don't have a chance without weapon or armor.

In Minecraft a normal zombie at normal Difficulty makes 3 points damage, but even with an iron armor you still get 1.38. 

That's why I suggest to use a float number instead of integer because integer is not accurate enough

**Comment by Ferk** [comment]
> That's why I suggest to use a float number instead of integer because integer is not accurate enough

Are you certain that this would already solve the issue? To me it looks like the mods need to be rebalanced, either the armor or some of the mobs (possibly both).

If you want to increase the precision you can multiply the health of the player by 10:

    player:set_properties({hp_max = 200})

then multiply the damage and health of all monsters by 10 as well, this would be equivalent to adding one decimal digit more of precision.

It looks like the wooden armor is too protective. Can we get the actual numbers for armor groups set by the 3d armor mod? I still think this should be discussed in the mod first, only bring it to the engine attention if you are certain that it's not a problem of the mod. The team working in the engine is not very big.

**Comment by MarkuBu** [comment]
> Are you certain that this would already solve the issue? 

I'm pretty sure, as long as the damage is always a float and only the visual representation on the screen is "rounded".

> If you want to increase the precision you can multiply the health of the player by 10 (the default is 20 "heart halves"):

That would be also possible, but float would still be more precise. I could multiply by 100, but then if will be ridiculous.  Even multiply by 10 would only be a workaround and not a solution. 

> It looks like the wooden armor is too protective.

No, with the correct calculation it would still make 80% of the damage. But because each decimal is rounded down it does no damage.

**Comment by Ferk** [comment]
> because each decimal is rounded down it does no damage.

Does it really do no damage at all? as in zero?

The difference is likely to be only half a heart. with proper rounding, and even less of a difference if instead of rounding it used the exact float number.

I can try and see if I can test this when I have the time, build a version with proper rounding and see if there's any real difference.

**Comment by MarkuBu** [comment]
> Does it really do no damage at all? as in zero?

I didn't look at the current values of armor and mob strength. I have download the latest versions of mobs redo and 3d armor 3 days ago and use the default settings. I made a new game and gave me a steel armor without shield. I was able to stand within multiple sand and dirt mobs without getting any damage. ZERO. Only spiders made damage

**Comment by MarkuBu** [comment]
And btw: with proper rounding the problem would just be reduced, not solved. You still can get 0 damage, except you say that 0.5 is the minimum

**Comment by MarkuBu** [comment]
To be honest, even if you round properly to an accuracy of 0.5 you have a problem because each value results in a full number. Even if you only would get 0.1 damage you can choose between 0 (witch is not wanted) and 1 (witch is too much). If you still get half a heart damage, even if you are attacked by a sand monster and you have a full diamond armor, then is something wrong

You need to calculate internally with float or you must multiply internally everything with 10, or better 100. To set hp_max to 200 is a dirty hack

**Comment by Ekdohibs** [comment]
Note that another way to inflict `x` damage is to inflict always `floor(x)` damage, plus one additional damage with probability `x-floor(x)`, which inflicts `x` expected damage.

**Comment by Ekdohibs** [comment]
Yes, you always get 0 or 1, but after, say, 20 attacks with strength 0.1, you get on average 2 damage.

**Comment by MarkuBu** [comment]
So then you have to store the damage somewhere in a float or in an integer multiplied by 10 or 100. Or did I miss something here?

**Comment by MarkuBu** [comment]
I replaced the s16 damage with float in tools.cpp and tools.h. Didn't get any error or warning message so far while compiling and game is running

The behavior of the game changed. Even with an iron armor sand monster can make damage now, but I assume that somewhere else damage is still handled as integer. Need to find this too. Can someone give me a hint?

**Comment by paramat** [comment]
Max health is now settable as a player object property.

**Comment by MarkuBu** [comment]
It's not about max value, it's about precision. 

Is it documented somewhere? Is it `set_hp()`?

**Comment by MarkuBu** [comment]
Ok, I think it is hp_max. But what happens if I set it to 200 for example? Do I still have 10 hearts and each heart stands for 20 instead of 2 hitpoints?

**Comment by Ferk** [comment]
I indicated a few comments above how:

    player:set_properties({hp_max = 200})

All documentation for the API is in [lua_api.txt](https://github.com/minetest/minetest/blob/master/doc/lua_api.txt), just search for hp_max.

**Comment by MarkuBu** [comment]
> I believe if you increase it, the number of hearts in the health bar increase as well unless you use a mod that uses progress bars.

That's the point and not the answer for my issue

**Comment by Ferk** [comment]
> That's the point and not the answer for my issue

It was meant as an answer to your question about the 10 hearts, not a solution to the issue.

In my opinion, if I were to make a change I would do something like in "Zelda: breath of the wild". I've been playing it lately and it also uses an integer damage system that is decent. But instead of having half hearts as the unit, it uses quarters of heart (so in Minetest that would involve doubling the default hp_max to 40, which gives 10 hearts that is a decent health meter in Zelda:BotW as well). Then have the engine always inflict a minimum damage of 1, so everything that hurts always does at minimum 1/4 heart damage, it's small enough. You can still be invulnerable to a type of damage by having its armor_group to zero, but any other armor value would inflict at least a minimum. Even when you are completely maxed out in BotW you still will get 1/4 heart damage from the weakest hit.

**Comment by lhofhansl** [comment]
I think changing the damage accumulator (`damage`) to a float is good change. It's not the final change, but 

The _final_ result is still converted to an s16 (see `HitParams` struct), but we won't round each part on the way.

Let's make that change now. Then we can also go back and fix the damage system.

**Comment by lhofhansl** [comment]
Also, seeing messages like these:
"Player XYZ punched by LuaEntitySAO at (-166.541,21.5,-142.93), damage 0 HP"
with just wooden armor on. These are Sandmonsters.

**Comment by MarkuBu** [comment]
I'm still confused that it is possible to change the type of a variable without any compiler warning

**Comment by lhofhansl** [comment]
It'll just convert the float to an integer upon return.

**Comment by Ferk** [comment]
I guess the warnings for conversion ``-Wconversion`` are not enabled.
I'd probably be better to make the rounding explicit when setting the HitParams, than relying on whatever casting the compiler does. Maybe it's just doing a "floor".

**Comment by lhofhansl** [comment]
I do agree that the current state is ... "nonsense". With simple wooden armor you can stand inside an army of sandmonsters or zombies and absolutely nothing will happen to you.

Either it needs to be scaled by (say) a factor of 10, or converted all to float. I think I'd prefer the latter.

**Comment by MarkuBu** [comment]
Wear only happens when a player gets fall damage. But I think that is a mod issue

**Comment by lhofhansl** [comment]
Shouldn't the armor wear off? Anyway... Different discussion :)

**Comment by jastevenson303** [comment]
@lhofhansl no, it's the same discussion : )
https://github.com/minetest/minetest/issues/6773#issuecomment-351060508

**Comment by MarkuBu** [comment]
But it is a 3d armor issue. Armor wear is working, but only for fall damage and not on punch damage from mobs. I don't know if armor get's wear if it is punched by another player

**Comment by lhofhansl** [comment]
So what should we do here? I surveyed the code a bit. Changing the damaging system from int to float has wide implications including a protocol change. If we wanted to tackle that it should probably be before 0.5.

**Comment by MarkuBu** [comment]
The thing is: the damage system doesn't really work well with int. Even if health is doubled and uses quarter hearts it is still not very precise.

You can multiply by 10, but that should be transparent. The player should still have 20hp and not 200

I don't know how many code has to be changed, but would give it a try

**Comment by Ferk** [comment]
I dont think damage needs to be very precise, it´s more important for it to be consistent than precise.
I´d rather have an enemy make always a minimum 1 point of damage regardless of armor than having it sometimes deal zero damage (because the hearts cannot represent 0.1 values) and other times a complete 1 point. It would become harder to know how many hits you can take.

**Comment by MarkuBu** [comment]
BotW has a less complex damage system.

However, with a random system it will also be hard to know how many hits you can get

A bit off topic: how can I override the built-in damage system? Will try to implement this in Lua to test my idea

**Comment by Ferk** [comment]
Less complex in which sense? it has more armor levels and specific resistances by damage type (fire/ice/electric). The fact that health is integer does not prevent you from having a rich system.

I agree that a random system is hard to know, that's why I was saying that a float system won't be very different from that.

I don´t think you can really override the builtin damage system in Lua. You could disable damage and use a custom player property instead of health to manage it from Lua and modify the mobs to decrease that property, making the player die when that property reaches zero. Normal weapons won't work, though.

**Comment by lhofhansl** [comment]
The current integer stuff is unintuitive. Perhaps it can all simply be scale by 100 (and thus all expressed in %).

**Comment by MarkuBu** [comment]
> You could disable damage and use a custom player property instead of health to manage it from Lua and modify the mobs to decrease that property, making the player die when that property reaches zero. Normal weapons won't work, though.

I know that I can disable everything with `immortal`. Will try to do something after Christmas

**Comment by MarkuBu** [comment]
It seems that fall damage is handled independent from the other damage system. But should be part of it and configurable. 

Currently it seems to be impossible to add something like status effects (fall damage protection) or change fall height. 

Fall damage is hard coded and can only disabled if you disable damage in the main menu

**Comment by MarkuBu** [comment]
Update the first comment. And now merry Christmas for all of you, wherever you live

**Comment by MarkuBu** [comment]
Just a thought: we know the size of the Minecraft hitbox, but is this identical with the collision box? Do Minecraft entities have some kind of collision box at all?

Because there can be multiple mobs at the same position.  If they have full size collision boxes this shouldn't be possible. Probably the only have a collision line in the center?

**Comment by stujones11** [comment]
The 3d_armor mod was originally based on MTG and is mostly geared towards pvp since the game has no mobs by default. Though I did add level coefficients to help address this problem, it's still not an complete fix.

The problem I have, as others have pointed out, is the lack of precision and the reason I added the 'healing' properties to armor. @Ekdohibs has an interesting suggestion which is effectively the opposite of this which might be worth looking into.

> Max health is now settable as a player object property.

This is indeed a step in the right direction, however, it needs to be implemented by the game. I do not think it's right that the armor mod should interfere with this,

**Comment by raymoo** [comment]
I support the idea to just have the fractional part of damage have a random chance to deal 1 damage (roll a number from 0 to 1 and check if it's under the value).

**Comment by ThePython10110** [comment]
MineCraft uses floats for HP.... so it's possible. But I realize that changing this in Minetest would be a lot of work.

Just noticed it's been almost 6 years... whoops.

**Comment by Zughy** [comment]
I don't know why, but I always thought that Luanti used a float value. Then I noticed how my guns dealing 0.3 damage per hit resulted in not changing the enemy's health of an inch after a whole magazine unloaded on them.  And the use case of OP makes completely sense as well.  
I agree that internally multiplying by 10 is a dirty hack, sooo... @lhofhansl do you think this might happen sooner or later? It's pretty limiting having to deal with integers
